mclient/dspipe.8
.\"
.\"
.\"
.\"
.\"
.\"
.TH DSPIPE 8 "16 October 2009" "MIT SIPB"
.SH NAME
dspipe \- Pipe STDIN into a discuss meeting
.SH SYNOPSIS
.B dspipe
.I meeting_name
[ \-t
.I topic
]
[ \-s
.I signature
]
.SH DESCRIPTION
.I dspipe
is designed to be used on discuss servers to pipe STDIN into a
specified discuss meeting.  A topic (subject) for the transaction may
be specified, as well as a signature.  The user running dspipe must
have "answer", "status", and "write" permissions on the meeting in
question.

.SH FILES
~/.meetings	Data file listing meetings that you attend.

.SH AUTHORS
Stan Zanarotti, MIT Student Information Processing Board

.SH "SEE ALSO"
.PP
discuss(1)







debian/changelog
debathena-discuss (10.0.12-0debathena1) unstable; urgency=low

  * Stop using the obsolete AC_TYPE_SIGNAL. (Trac: #347)
  * Add missing manpages. (Trac: #266)

 -- Evan Broder <broder@mit.edu>  Sun, 15 Nov 2009 17:35:19 -0500

debathena-discuss (10.0.11-0debathena1) unstable; urgency=low

  * Move the single-letter checks in trn_expr.y:yylex to the bottom of the
    function so that they get caught after commands that start with those
    letters. (Trac: #346)

 -- Evan Broder <broder@mit.edu>  Sat, 29 Aug 2009 17:20:43 -0400

debathena-discuss (10.0.10-0debathena2) unstable; urgency=low

  * Depend on Kerberos-enabled zephyr instead of build-depending.

 -- Evan Broder <broder@mit.edu>  Mon, 27 Jul 2009 14:38:10 -0400

debathena-discuss (10.0.10-0debathena1) unstable; urgency=low

  * Patch discuss server to work without krb4. Patch by Mitch Berger.

 -- Evan Broder <broder@mit.edu>  Thu, 23 Jul 2009 21:00:55 -0400

debathena-discuss (10.0.9-0debathena1) unstable; urgency=low

  * Fix local/dumb authentication needed for mail delivery and local
    client usage. Patch by Mitch Berger. (Trac: #274)
  * Always build without krb4.

 -- Evan Broder <broder@mit.edu>  Wed, 22 Jul 2009 22:58:40 -0400

debathena-discuss (10.0.8-0debathena1) unstable; urgency=low

  * Add one-character aliases to discuss. Patch by Ben Kaduk. (Trac: #238)
  * Squash some lintian warnings by switching from ${Source-Version} to
    ${source:Version} and ${binary:Version}
  * Use HAVE_KRB4 and HAVE_KRB5 instead of KERBEROS and KERBEROS5.
  * Assorted changes to make discuss work with any combination of krb4 and
    krb5 (including neither).
  * Always build our discuss with krb5.

 -- Evan Broder <broder@mit.edu>  Mon, 18 May 2009 23:51:52 -0400

debathena-discuss (10.0.7-0debathena3) unstable; urgency=low

  * Configure the discuss inetd service.

 -- Evan Broder <broder@mit.edu>  Thu, 05 Mar 2009 15:17:20 -0500

debathena-discuss (10.0.7-0debathena2) unstable; urgency=low

  * Change DEB_AUTO_UPDATE_AUTOCONF to 2.50, not 1.
  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.
  * Clean up debian/copyright.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 14:31:41 -0500

debathena-discuss (10.0.7-0debathena1) unstable; urgency=low

  * Revert a small part of the OS X com_err patch from 10.0.6 in favor of
    changes to aclocal.m4
  * Remove hard-coded paths from discuss-mode, in favor of getting them
    from autoconf.
  * Remove debathena-machtype as a dependency, as it's no longer needed.

 -- Evan Broder <broder@mit.edu>  Fri, 09 Jan 2009 21:47:33 -0500

debathena-discuss (10.0.6-0debathena1) unstable; urgency=low

  [ Evan Broder ]
  * Add com_err support for delete and discuss on OS X.

  [ Anders Kaseorg ]
  * Depend emacs | emacsen instead of emacs22 | emacs21 | emacsen, so that
    the emacs metapackage is pulled in by default.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 08 Jan 2009 16:35:04 -0500

debathena-discuss (10.0.5-0debathena1) unstable; urgency=low

  * Fix some trivial typos.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 19 Aug 2008 16:35:07 -0400

debathena-discuss (10.0.4-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc on 64-bit platforms.
  * Fix date display in discuss on 64-bit platforms.

 -- Greg Hudson <ghudson@mit.edu>  Mon, 30 Jun 2008 12:50:56 -0400

debathena-discuss (10.0.3-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc (int/long mismatch in short_time).
  * Put edsc crash dumps in /tmp instead of /usr/tmp.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 19:39:38 -0400

debathena-discuss (10.0.2-0debathena1) unstable; urgency=low

  * Stop assuming that long is 32 bits.
  * Start assuming that int is 32 bits.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 16:29:35 -0400

debathena-discuss (10.0.1-0debathena1) unstable; urgency=low

  * Fix obsolete names for Kerberos library functions, which no longer
    worked on Mac OS X.  Patch from broder.

 -- Anders Kaseorg <andersk@mit.edu>  Sun, 06 Apr 2008 19:25:13 -0400

debathena-discuss (10.0.0-0debathena2) unstable; urgency=low

  * Actually use SERVERDIR, BINDIR, and INFODIR in config.h.sed.  (The
    Makefile already interpolates these variables.)

 -- Anders Kaseorg <andersk@mit.edu>  Sun, 06 Apr 2008 17:28:52 -0400

debathena-discuss (10.0.0-0debathena1) unstable; urgency=low

  * Merged quilt patches into mainline Athena tree

 -- Shreevatsa R <vatsa@mit.edu>  Sun,  6 Apr 2008 15:54:22 -0400

debathena-discuss (9.4.41-0debathena2) unstable; urgency=low

  * Add emacs22 as a default alternative to emacs21 | emacsen.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 29 Oct 2007 21:40:29 -0400

debathena-discuss (9.4.41-0debathena1) unstable; urgency=low

  * New upstream release supporting Kerberos 5.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 23 Aug 2007 14:28:28 -0400

debathena-discuss (9.4.30-0debathena3) unstable; urgency=low

  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:33:41 -0400

debathena-discuss (9.4.30-0debathena2) unstable; urgency=low

  * Move edsc to debathena-discuss package.
  * Don't link libreadline, because libss is stupid and tries to do so at
    runtime.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 03:37:48 -0400

debathena-discuss (9.4.30-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 30 Jan 2007 22:57:54 -0500


mclient/dspipe.c
/*
 *
 *	Copyright (C) 1988, 1989 by the Massachusetts Institute of Technology
 *    	Developed by the MIT Student Information Processing Board (SIPB).
 *    	For copying information, see the file mit-copyright.h in this release.
 *
 */
/*
 *
 * dspipe.c -- Program to pipe stdin into a Discuss meeting.
 *
 */

#include <discuss/tfile.h>
#include <discuss/interface.h>
#include <discuss/dsc_et.h>
#include <rpc.h>
#include <sys/file.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <unistd.h>

tfile unix_tfile();
char *mktemp();

#ifndef	lint
static char rcsid[] = "$Id: dspipe.c,v 1.11 2002-04-06 15:18:51 zacheiss Exp $";
#endif

main (argc,argv)
int argc;
char **argv;
{
	char *usr_mtg = "";
	char *subject = "";
	char *signature = "";
	char mtg_name [100],host[70];
	int result;
	int fatal_err;
	trn_nums result_trn;
	char filename [40], buffer [512];
	tfile tf;
	int i,nread,d,exit_code;

#if defined(__APPLE__) && defined(__MACH__)
	add_error_table(&et_dsc_error_table);
#else
	initialize_dsc_error_table();
#endif
	for (i = 1; i < argc; i++) {
		if (*argv[i] == '-')
			switch (argv[i][1]) {
			case 'd':
			case 'a':
				continue;
				
			case 't':
				if (++i < argc)
					subject = argv[i];
				continue;

			case 's':
				if (++i < argc)
					signature = argv[i];
				continue;

			default:
				goto lusage;
			}
		if (*usr_mtg != '\0')
			goto lusage; /* only one meeting name */
		
		usr_mtg = argv[i];
	}
	
	if (*usr_mtg == '\0')
		goto lusage;


	(void) strcpy (filename, "/tmp/DSXXXXXX");
	(void) mktemp (filename);

	d = open (filename, O_RDWR | O_CREAT, 0700);
	if (d < 0) {
		(void) fprintf (stderr, "Cannot open temp file\n");
		exit (1);
	}

	for (;;) {
		nread = read (0, buffer, 512);
		if (nread < 0) {
			(void) fprintf (stderr, "Cannot read stdin\n");
		}
		write (d, buffer, nread);
		if (nread == 0)
			break;
	}

	lseek (d, 0, SEEK_SET);	/* rewind temp */

	tf = unix_tfile (d);

	strcpy (host, "discuss@");
	resolve_mtg (usr_mtg, &host[8], mtg_name);

	init_rpc();
	set_module (host, &fatal_err, &result);
	if (fatal_err) {
	     (void) fprintf (stderr, "%s\n", error_message(result));
	     exit(1);
	}
	if (result) {
	     (void) fprintf (stderr, "Warning: %s\n", error_message(result));
	}

	if (signature[0] == '\0') {
	     add_trn (mtg_name, tf, subject, 0, &result_trn, &result);
	} else {
	     if (get_server_version() < SERVER_2)
		  add_trn (mtg_name, tf, subject, 0, &result_trn, &result);
	     else
		  add_trn2(mtg_name, tf, subject, signature, 0, &result_trn, &result);
	}
		  
	if (result != 0) {
		(void) fprintf (stderr, "%s\n", error_message(result));
		exit_code = 1;
		goto bye;
	}

	(void) printf ("Transaction [%04d] added to %s\n", result_trn,
		       mtg_name);
	exit_code = 0;

 bye:
	(void) close (d);
	tdestroy (tf);
	(void) unlink (filename);

	term_rpc();
	exit (exit_code);

 lusage:
	(void) fprintf (stderr, "Usage: dspipe {mtgname} -t {topic} -s {signature}\n");
	exit (1);
}

/*
 *
 * resolve_mtg:  Procedure to resolve a user meeting name into its host
 * 	         an pathname.
 *
 */
resolve_mtg (usr_string, host, mtg_name)
char *usr_string,*host,*mtg_name;
{
	char *colon;
	int host_len;

	colon = strchr (usr_string, ':');

	if (colon == 0) {
		(void) strcpy (mtg_name, usr_string);
		gethostname (host, 50);
		return;
	}

	host_len = colon - usr_string;
        memmove (host, usr_string, host_len);
	host [host_len] = '\0';
	(void) strcpy (mtg_name, colon+1);
	return;
}
